# Solutions to Lesson 2 Exercises

For each exercise, the solutions below show one possible way of solving it, but you might have used a different approach, and that's great! There is almost always more than one way to solve any particular problem in Python.

## Initial Setup

Since this notebook is in the `solutions` sub-folder, use the  magic command `%cd` to go up one folder to the main project folder to keep the file paths the same as in the lessons:

In [1]:
%cd ..

C:\Users\jenfl\Projects\eoas-python


Import `pandas` library:

In [2]:
import pandas

# Exercise 2

**a)** Read the file `data/gapminder_world_data_2018.csv` into a new DataFrame `world` and display the first 10 rows.

In [3]:
world = pandas.read_csv('data/gapminder_world_data_2018.csv')
world.head()

country  population    region                       sub_region  \
0          Afghanistan    36400000      Asia                    Southern Asia   
1              Albania     2930000    Europe                  Southern Europe   
2              Algeria    42000000    Africa                  Northern Africa   
3               Angola    30800000    Africa               Sub-Saharan Africa   
4  Antigua and Barbuda      103000  Americas  Latin America and the Caribbean   

   income_group  life_expectancy  gdp_per_capita  children_per_woman  \
0           Low             58.7            1870                4.33   
1  Upper middle             78.0           12400                1.71   
2  Upper middle             77.9           13700                2.64   
3  Lower middle             65.2            5850                5.55   
4          High             77.6           21000                2.03   

   child_mortality  pop_density  
0            65.90         55.7  
1            12.90        107.0  
2            23.10         17.6  
3            81.60         24.7  
4             7.89        234.0

**b)** How many rows and columns does `world` have?

In [4]:
world.shape

(178, 10)

178 rows and 10 columns

**c)** Display the names and data types of each column.

In [5]:
world.dtypes

country                object
population              int64
region                 object
sub_region             object
income_group           object
life_expectancy       float64
gdp_per_capita          int64
children_per_woman    float64
child_mortality       float64
pop_density           float64
dtype: object

**d)** Display summary statistics with the `describe` method. What are the lowest and highest populations? How about lowest/highest population densities? Any guesses which countries these might be? (We'll find out the answer in Lesson 4!)

In [6]:
world.describe()

population  life_expectancy  gdp_per_capita  children_per_woman  \
count  1.780000e+02       178.000000      178.000000          178.000000   
mean   4.266072e+07        72.653371    17892.808989            2.746573   
std    1.511449e+08         7.339901    19330.555705            1.294496   
min    9.520000e+04        51.100000      629.000000            1.230000   
25%    3.207500e+06        67.100000     3572.500000            1.750000   
50%    9.910000e+06        74.050000    12200.000000            2.265000   
75%    3.170000e+07        78.175000    25350.000000            3.647500   
max    1.420000e+09        84.200000   121000.000000            7.130000   

       child_mortality  pop_density  
count       177.000000   178.000000  
mean         29.767401   201.118202  
std          29.613831   660.593969  
min           1.950000     2.010000  
25%           7.410000    30.850000  
50%          16.700000    81.000000  
75%          49.700000   151.000000  
max         126.000000  8270.000000

- The lowest and highest populations are 95200 and 1.42 billion, respectively.
- The lowest and highest population densities are 2.01 and 8270 people per km$^2$, respectively.

### Bonus exercises

**e) Data wrangling - dealing with header rows**

The file `data/raw/weather_YVR_1938.csv` contains the daily weather data for 1938, in the original format downloaded from Environment Canada. Open this file in the JupyterLab CSV viewer to see what it looks like.

> Note that the CSV viewer isn't able to parse the data correctly because of the extra header rows at the beginning.

- Now try reading the file into your notebook with `pandas.read_csv` and see what happens.

If you look at the documentation for `pandas.read_csv`, you'll see a `skiprows` input buried amongst a few dozen other inputs for this function. This input tells `read_csv` how many rows to skip at the beginning of the file. 
- Try reading `data/raw/weather_YVR_1938.csv` again, but this time using a value of `24` for the `skiprows` keyword argument, and display the first 5 rows of the resulting DataFrame.

In [7]:
weather_1938 = pandas.read_csv('data/raw/weather_YVR_1938.csv', skiprows=24)
weather_1938.head()

Date/Time  Year  Month  Day  Data Quality  Max Temp (°C) Max Temp Flag  \
0  1938-01-01  1938      1    1           NaN            9.4           NaN   
1  1938-01-02  1938      1    2           NaN            7.2           NaN   
2  1938-01-03  1938      1    3           NaN            7.2           NaN   
3  1938-01-04  1938      1    4           NaN            7.2           NaN   
4  1938-01-05  1938      1    5           NaN            7.2           NaN   

   Min Temp (°C)  Min Temp Flag  Mean Temp (°C)          ...           \
0           -0.6            NaN             4.4          ...            
1            1.7            NaN             4.5          ...            
2           -3.9            NaN             1.7          ...            
3           -2.8            NaN             2.2          ...            
4           -2.8            NaN             2.2          ...            

  Total Snow (cm)  Total Snow Flag Total Precip (mm)  Total Precip Flag  \
0             NaN                M               0.3                NaN   
1             NaN                M               0.5                NaN   
2             0.0              NaN               0.0                NaN   
3             0.0              NaN               0.0                NaN   
4             0.0              NaN               0.0                NaN   

  Snow on Grnd (cm)  Snow on Grnd Flag Dir of Max Gust (10s deg)  \
0               NaN                NaN                       NaN   
1               NaN                NaN                       NaN   
2               NaN                NaN                       NaN   
3               NaN                NaN                       NaN   
4               NaN                NaN                       NaN   

   Dir of Max Gust Flag Spd of Max Gust (km/h)  Spd of Max Gust Flag  
0                   NaN                    NaN                   NaN  
1                   NaN                    NaN                   NaN  
2                   NaN                    NaN                   NaN  
3                   NaN                    NaN                   NaN  
4                   NaN                    NaN                   NaN  

[5 rows x 27 columns]

**f) Importing a library from a `.py` file**

In the workshop folder, you'll see a file called `ecweather.py`. It is a Python *module*, which is a library contained in a single `.py` file (as opposed to a package, which is multiple `.py` files bundled together).

- Double-click on this file in the Files Sidebar of JupyterLab to view it in the text editor. You'll see that it contains two functions: `welcome` and `download_daily_data`.

You can import a library from a local `.py` file with the same syntax as any other library. The library name is just the file name minus the `.py` extension, so to import this library the syntax is:
```python
import ecweather
```

- Import `ecweather` into your notebook, and call the function `ecweather.welcome()` to test it. If everything worked ok, it should print a welcome message.

> We'll use the other function `download_daily_data` in a later bonus exercise!

In [8]:
import ecweather

In [9]:
ecweather.welcome()

Hello! Thank you for importing the ecweather library!


**g) Library nicknames** 

When you import a library you can give it a nickname. For example, the `pandas` library is commonly imported with the syntax:
```python
import pandas as pd
```
This gives a nickname `pd` to the library, which is used instead of the full name to save a bit of typing (we scientists tend to be very lazy when it comes to coding). So instead of `pandas.read_csv`, we would type `pd.read_csv`, and so on for any other functions or other items in the library.

- Import the `numpy` library and give it a nickname `np`.
- Use the `mean` function from this library to compute the mean of the list `[5, 6, 7, 8, 9, 10]`.

In [10]:
import numpy as np

In [11]:
np.mean([5, 6, 7, 8, 9, 10])

7.5